### PR TITLE
Configurable path prefix to run PromDash behind a proxy

### PR DESCRIPTION
**Pull request by quinox**
Same as https://github.com/prometheus/prometheus/pull/614 but then for PromDash.

This pull request should be closely reviewed, I have no experience with Rails. All feedback greatly appreciated.

According to `rake routes` everything is covered. I am currently running PromDash on this code behind a proxy without any issues.

* I had to delete the `tmp/` folder to flush some cached files. I have no idea if this is always required.
* `graph_template.html.erb` and `pie_template.html.erb` contain an `<a href="/{{dashboard.slug}}">` but I have no idea how those tags are used

### CODE REVIEW

**Comment by juliusv** [comment]
@quinox Sorry, I'll probably only get to this in >1 week due to vacation. Or maybe @stuartnelson3 wants to take a look (if yes, please verify that all instances of using the prefix actually work). In any case, PromDash is a slow-moving target at the moment, so I would expect everything to still apply without conflict in a week.

**Comment by quinox** [comment]
No problem, my own system is working fine so there's no rush. The patch is rather straight forward, so easy to fix if it goes into conflict.

**Comment by juliusv** [comment]
Getting to this now. To be sure that things still apply cleanly, could you rebase ontop of current master?

**Comment by quinox** [comment]
Squashed and rebased.

**Comment by juliusv** [comment]
Thanks, taking a look now.

**Comment by juliusv** [comment]
Great, most of this seems to be working fine, but I still found three areas where it's using the unprefixed path and thus gets a `404` / breaking functionality:

- "Link to Graph" button on any widget (`/w`)
- "Clone Widget" (`/directories.json`)
- "Create permalink to current dashboard state" (`/permalink`)

File: app/assets/javascripts/angular/controllers/dashboard_ctrl.js.erb [hunk 1]
```diff
@@ -189,7 +189,7 @@ angular.module("Prometheus.controllers")
 
   $scope.queryDirectory = function() {
     $scope.directoryForClone.id = $scope.directoryForClone.id || "unassigned";
-    $http.get('/directories/' + $scope.directoryForClone.id).then(function(payload) {
+    $http.get('<%= Rails.configuration.path_prefix %>directories/' + $scope.directoryForClone.id).then(function(payload) {
```
Comment (juliusv) [inline app/assets/javascripts/angular/controllers/dashboard_ctrl.js.erb]: The prefix usages in this file all need a slash after the path prefix to work.

File: app/assets/javascripts/angular/services/annotation_refresher.js.erb [hunk 2]
```diff
@@ -13,7 +13,7 @@ angular.module("Prometheus.services").factory('AnnotationRefresher', ["$http", "
       var until = Math.floor(graph.endTime || Date.now()) / 1000;
 
       tags.forEach(function(t) {
-        $http.get('/annotations', {
+        $http.get('<%= Rails.configuration.path_prefix %>annotations', {
```
Comment (juliusv) [inline app/assets/javascripts/angular/services/annotation_refresher.js.erb]: Same here (slash separator missing).

File: app/assets/javascripts/angular/services/widget_link_helper.js.erb [hunk 3]
```diff
@@ -1,7 +1,7 @@
 angular.module("Prometheus.services").factory('WidgetLinkHelper', ["$http", "InputHighlighter", function($http, InputHighlighter) {
   return {
     createLink: function(data) {
-      this.promise = $http.post('/w', data);
+      this.promise = $http.post('<%= Rails.configuration.path_prefix %>w', data);
```
Comment (juliusv) [inline app/assets/javascripts/angular/services/widget_link_helper.js.erb]: Dito here and below.

**Comment by juliusv** [comment]
Hmm, actually, it does work when I end the prefix itself with a slash (thought at first it didn't, but must have done something wrong). I wonder, could we auto-sanitize the prefix somehow so that it always contains a slash, no matter what the config provides?

**Comment by juliusv** [comment]
Ah yeah, the reason I first thought it didn't work at all even with a slash in the prefix were the cached temp files, I think.

**Comment by juliusv** [comment]
But definitely, the dashboard permalinking (the little link button at the top in the global settings) needs a prefix added in these two places:

- https://github.com/prometheus/promdash/blob/master/app/assets/javascripts/angular/controllers/dashboard_ctrl.js#L35
- https://github.com/prometheus/promdash/blob/master/app/controllers/dashboards_controller.rb#L51

**Comment by quinox** [comment]
I sanitized path_prefix inside `environment.rb`, seems like a good spot (but again, I don't know any Rails).

**Comment by quinox** [comment]
Note to self: Remove settings from `environments/*.rb`, set them inside `config/application.rb` based on ENV and auto-update `config.path_prefix` (when needed)

**Comment by quinox** [comment]
Implemented the move from config file to ENV variable.

In my research I came across an existing ENV variable named `RAILS_RELATIVE_URL_ROOT`. The documentation about this variable talks about deploying to another directory which this feature isn't about (`path_prefix` is purely for routing). When I used it it also messed up serving static files so I decided to use a different variable name.

**Comment by juliusv** [comment]
Excellent, everything I tried worked! :+1: Could you squash it? Then I'll merge.

**Comment by quinox** [comment]
It's all prepped.

**Comment by juliusv** [comment]
Thanks!

**Pull request merged**